tsl/src/compression/compression.c

```diff
@@ -2319,19 +2323,47 @@ decompress_batches(RowDecompressor *decompressor, ScanKeyData *scankeys, int num
 									&tmfd,
 									false);
 
-		if (result == TM_Updated || result == TM_Deleted)
-		{
-			if (IsolationUsesXactSnapshot())
-				ereport(ERROR,
-						(errcode(ERRCODE_T_R_SERIALIZATION_FAILURE),
-						 errmsg("could not serialize access due to concurrent update")));
-			return false;
-		}
-		if (result == TM_Invisible)
+		switch (result)
 		{
-			elog(ERROR, "attempted to lock invisible tuple");
-			return false;
+			case TM_Deleted:
+			{
+				if (IsolationUsesXactSnapshot())
+				{
+					ereport(ERROR,
+							(errcode(ERRCODE_T_R_SERIALIZATION_FAILURE),
+							 errmsg("could not serialize access due to concurrent update")));
+					table_endscan(heapScan);
+					return false;
+				}
+				continue;
+			}
+			break;
+			case TM_BeingModified:
+			case TM_SelfModified:
+				Assert(false);
+				break;
+			/*
+			 * DML on compressed hypertable and a direct UPDATE on compressed chunk
+			 * can cause executor to return TM_Updated.
+			 */
+			case TM_Updated:
+			{
+				elog(ERROR, "tuple concurrently updated");
+				table_endscan(heapScan);
+				return false;
+			}
+			break;
+			case TM_Invisible:
+			{
+				elog(ERROR, "attempted to lock invisible tuple");
+				table_endscan(heapScan);
+				return false;
+			}
+			break;
+			default:
+				break;
```

Comment:
this is wrong; the `default` should be the error; and the other which are letting things thru